src/cudamatrix/cu-kernels.cu

```diff
@@ -773,6 +773,142 @@ static void _add_mat_blocks_trans(Real alpha, const Real* src,
     }
 }
 
+template<typename Real>
+__global__
+static void _max_mat_blocks(const Real *src, Real *dst, Real *index_max_,
+                            const int32_cuda pool_t_size_,                            
+                            const int32_cuda pool_h_size_,
+                            const int32_cuda pool_f_size_,
+                            const int32_cuda pool_t_step_,
+                            const int32_cuda pool_h_step_,
+                            const int32_cuda pool_f_step_,
+                            const int32_cuda input_h_dim_,
+                            const int32_cuda input_f_dim_) {
+  int32_cuda i = blockIdx.x * blockDim.x + threadIdx.x;
+  int32_cuda j = blockIdx.y * blockDim.y + threadIdx.y;
+  int32_cuda k = blockIdx.z * blcokDim.z + threadIdx.z;
+  int32_cuda num_pools_h = 1 + (input_h_dim_ - pool_h_size_) / pool_h_step_;
+  int32_cuda num_pools_f = 1 + (input_f_dim_ - pool_f_size_) / pool_f_step_;
+
+  // initialize the temporary maximum value and its index in each pool
+  int32_cuda max_row = i * pool_t_step_; 
+  int32_cuda max_col = j * pool_h_step_ * input_f_dim_ + k * pool_f_step_;
+  int32_cuda max_value = src[max_row][max_col];
+
+  // loop over all the elements in each pool to find the maximum one,
+  // and record its index.
+
+  for (int32_cuda t = 0; t < pool_t_size_; t++) {
```

Comment:
Do you mean I should set all the stride as 1?
the stride is represented by pool_{t,h,f}_step. I would add some description.

Reply:
pool_{t,h,f}_step is the stride between blocks, but as far as I found out the stride within block is 1.

Reply:
Do you mean it is possible that I need to get the maximum of part of the block rather than the whole block?
Like getting 5 from block [1,6,5,7,1] with within_block_stride as 2.

Reply:
I mean the rows of block can be something like [1 5 9]!

Reply:
I see what you mean, I would fix this! Thank you!
By the way, is this kind of stride only apply for rows? would the column also be like [1 5 9]?